File: 01_Arrays_and_Hashing/07_Product_of_Array_Except_Self.py
# TIME - O(n) : Two pass solution
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        ans = [1] * len(nums)

        prefix = 1
        postfix = 1

        for i in range(len(nums)):
            ans[i] = prefix
            prefix *= nums[i]

        for i in range(len(nums) - 1, -1, -1):
            ans[i] *= postfix
            postfix *= nums[i]

        return ans

File: 01_Arrays_and_Hashing/07_Product_of_Array_Except_Self.java
// TIME - O(n) : Two pass solution
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length, prefix = 1, postfix = 1;
        int[] ans = new int[n];

        for (int i = 0; i < n; i++) {
            ans[i] = prefix;
            prefix *= nums[i];
        }

        for (int i = n - 1; i >= 0; i--) {
            ans[i] *= postfix;
            postfix *= nums[i];
        }

        return ans;
    }
}